spec/mem/tree_borrows/tree.md

```diff
@@ -0,0 +1,145 @@
+When a node is accessed, each node in the tree can be divided into two disjoint sets: *child nodes* and *foreign nodes*, based on its relative position to the accessed node.
+The *child* set includes the node itself and all its descendants, while the *foreign* set contains all other nodes.
+
+```rust
+enum NodeRelation {
+    Child, 
+    Foreign,
+}
+```
+
+The state transition depends on both the node relation and the type of access operation.
+
+```rust
+enum AccessKind {
+    Read, 
+    Write,
+}
+```
+
+Then we implement how to actually update and check states in a Tree during each memory access.
+
+```rust
+impl Node {
+    /// Perform state transition on all locations of `self`.
+    /// The `node_relation` indicates whether the accessed node is a child or foreign
+    /// *from the perspective of `self`*.
+    fn transition(
+        &mut self, 
+        node_relation: NodeRelation,
+        access_kind: AccessKind,
+        offset_in_alloc: Size,
+        size: Size
+    ) -> Result {
+        let offset_start = offset_in_alloc.bytes();
+        for offset in offset_start..offset_start + size.bytes() {
+            self.location_states.mutate_at(offset, |location_state|{
+                location_state.transition(access_kind, node_relation)
+            })?;
+        }
+
+        ret(())
+    }
+
+    /// Recusively do state transition on `self` node and all its descendants.
+    /// `path` is the path from `self` to the accessed node.
+    /// `path` is None when the accessed node is not a descendant of `self`.
+    /// The `is_ancestor` indicates whether `self` is an ancestor of the accessed node.
+    ///
+    /// This method will throw UBs, representing the undefined behavior captured by Tree Borrows.
+    fn access(
+        &mut self,
+        path: Option<Path>, // self -> child
+        access_kind: AccessKind,
+        offset_in_alloc: Size,
+        size: Size,
+    ) -> Result {
+        // Indicates whether the accessed node is a child or foreign
+        // *from the perspective of `self`*.
+        //
+        // If `self` is an ancestor of the accessed node, the accessed node is a child of `self`.
+        // Otherwise, the accessed node is a `foreign` of `self`
+        let node_relation = if path.is_some() { NodeRelation::Child } else { NodeRelation::Foreign };
+
+        // perform state transition on `self`
+        self.transition(node_relation, access_kind, offset_in_alloc, size)?;
+
+        for child_id in Int::ZERO..self.children.len() {
+            // If the path starts with this child, do a child access with the path shortened by the first element.
+            // Otherwise, do a child access with the path being None (i.e., child is not an ancestor of accessed node)
+            let sub_path = if let Some(path) = path {
+                match path.first() {
+                    Some(first) if first == child_id => Some(path.subslice_with_length(Int::ONE, path.len() - Int::ONE)),
```

Comment:
This should also `split_first` once that exists:
`if let Some((head, tail)) = path.and_then(|p| p.split_first()) { ... }`